# Data Consistancy Checks

## Importing libraries

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
#creating directory path
path = r'C:\Users\luke_\Desktop\CF Datacourse\Python_exercise4\09-2024 Instacart Basket Analysis\02_data'

In [3]:
# Importing products.csv
df_prods = pd.read_csv(os.path.join(path,'original_data', 'products.csv'), index_col = False)

In [5]:
# Importing orders_wrangled.csv
df_ords = pd.read_csv(os.path.join(path,'prepared_data', 'orders_wrangled.csv'), index_col = False)

# Checking for missing values

In [12]:
# Checking df_prods for null values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
# creating a subset dataframe to isolate null product names from df_prodsdf_
df_nan=df_prods[df_prods['product_name'].isnull() == True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [16]:
# Checking rows and columns of df_prods
df_prods.shape

(49693, 5)

# Creating a mixed column in a new dataframe for practice

In [32]:
# creating a mix type dataframe
df_test = pd.DataFrame()

In [33]:
#creating a mixed column
df_test['mix'] = ['a', 'b', 1, True]

In [34]:
# Checking df_test
df_test.head()

mix
0     a
1     b
2     1
3  True

## Script to check for mixed columns in the dataframe.

In [36]:
#Because .applymap has been deprecated, we need to use .map. Careerfoundry should update their notes!
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [37]:
# correcting the mix column
df_test['mix'] = df_test['mix'].astype('str')

In [19]:
# Creating new dataframe from df_prods without null values
df_prods_clean = df_prods[df_prods['product_name']. isnull() ==False]

In [20]:
#checking rows and columns of df_prods_clean
df_prods.shape

(49693, 5)

## Finding duplicates and a subset to check them in

In [21]:
#creating subset of full duplicates from df_prods_clean
df_dubs = df_prods_clean[df_prods_clean.duplicated()]

In [22]:
df_dubs

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [23]:
# creating dataframe without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [24]:
df_prods_clean_no_dups.shape

(49672, 5)

## exporting products_checked.csv

In [26]:
df_prods_clean_no_dups.to_csv(os.path.join(path, 'prepared_data', 'products_checked.csv'))

# Exercise 4.5

## 2.Run the df.describe() function on your df_ords dataframe.

In [28]:
# Running the describe function on df_ords
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  time_of_order  days_since_prior_order  
count       3.421083e+06   3.421083e+06            3.214874e+06  
mean        2.776219e+00   1.345202e+01            1.111484e+01  
std         2.046829e+00   4.226088e+00            9.206737e+00  
min         0.000000e+00   0.000000e+00            0.000000e+00  
25%         1.000000e+00   1.000000e+01            4.000000e+00  
50%         3.000000e+00   1.300000e+01            7.000000e+00  
75%         5.000000e+00   1.600000e+01            1.500000e+01  
max         6.000000e+00   2.300000e+01            3.000000e+01

## There is an 'unnamed' column which I have tried to drop using  df_ords.drop(columns = ['Unnamed:0']). It seems this is a second Index column. This has not worked so I will ignore for now and carry on

## I have ignored the columns 'Unnamed', 'order_id', 'user_id' and 'order_number' as they are not numerical. 
## The 'order_day_of_week' has a min of 0 and a max of 6 with a mean of 3 as expected.
## The 'time_of_order' column we see floats. I would find out if this column records hours and minutes, or only 0-23.
## days since prior order also seems to be with in the expected range. I would say we have no reason to investigate further.

## 3. Check for mixed-type data in your df_ords dataframe.

In [38]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

## We did not get any returns and so we do not have any mixed types in the dataframe.

## 5.Run a check for missing values in your df_ords dataframe. 

In [39]:
# Checking df_ords for null values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_day_of_week              0
time_of_order                  0
days_since_prior_order    206209
dtype: int64

In [40]:
df_ords.shape

(3421083, 8)

## We see that there are no missing values exept for 206209 in 'days_since_prior_order'. As this is over 6% I will not drop the rows and rather speak to colleagues how best to approach this. Possibilities would be to ignore this or imputation.

## 6. Address the missing values using an appropriate method

## As the missing data is almost 6% we need to address this missing data. Idealy I would like to find out if there is a pattern to the missing data. Is it the same customer? is it always together with another variable, for instance, day of the week, or product? If there is no pattern I would imput the mean or median so as to ensure consistancy with further analysis.

## 7.Run a check for duplicate values in your df_ords data. 

In [42]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [44]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_day_of_week, time_of_order, days_since_prior_order]
Index: []

## We have not recieved any duplicates returned. We can assume that there are no duplicates in our df_ords Dataframe.

## Exporting orders_checked.csv

In [45]:
df_ords.to_csv(os.path.join(path, 'prepared_data', 'orders_checked.csv'))